libbeat/common/wait/timer.go

```diff
@@ -0,0 +1,124 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package wait
+
+import (
+	"math/rand"
+	"time"
+)
+
+// Timer represents a timer implementation.
+type Timer interface {
+	Reset(time.Duration)
+	Start()
+	Stop()
+	Wait() <-chan time.Time
+}
+
+// Waiter is the strategy to be used to find out the time to wait before making a call.
+type Waiter func() time.Duration
+
+// MinWaitAndJitter takes a minimal wait time and jitter and will return a duration that will be at
+// least equal of bigger than the initial wait time.
+func MinWaitAndJitter(d, jitter time.Duration) Waiter {
+	return func() time.Duration {
+		return d + time.Duration(rand.Int63n(int64(jitter)))
+	}
+}
+
+// RandomDelay takes a maximum duration and will return a random values which will be
+// max <= values > 0.
+func RandomDelay(max time.Duration) Waiter {
+	return func() time.Duration {
+		return time.Duration(int64(max))
+	}
+}
+
+// Fix waits for a fixed duration.
+func Fix(d time.Duration) Waiter {
+	return func() time.Duration {
+		return d
+	}
+}
+
+// PeriodicTimer uses a two Waiter strategy, the first will decide how much time to wait before
+// receive the initial tick and the other will be how much time to wait before doing the other calls,
+// this can be used to introduce more randomness in the frequency of the calls for outside system.
+// This can be useful if you want to better distribute calls that could affect the performance of
+// other system.
+type PeriodicTimer struct {
+	c           chan time.Time
+	resetOrDone chan time.Duration
+	initial     Waiter
+	periodic    Waiter
+	period      time.Duration
+}
+
+// NewPeriodicTimer returns a wait, allowing to wait for a minimum time and a random amount.
+func NewPeriodicTimer(initial, periodic Waiter) *PeriodicTimer {
+	jt := &PeriodicTimer{
+		c:           make(chan time.Time),
+		resetOrDone: make(chan time.Duration, 1),
+		period:      initial(),
+		periodic:    periodic,
+	}
+	return jt
+}
+
+// Start starts the timer.
+func (jt *PeriodicTimer) Start() {
+	go jt.startTimer()
+}
+
+func (jt *PeriodicTimer) startTimer() {
+	for {
+		select {
+		case reset, ok := <-jt.resetOrDone:
+			if !ok {
+				return
+			}
+			jt.period = reset
+		case <-time.After(jt.period):
+			jt.c <- time.Now()
+			jt.period = jt.periodic()
+		}
+	}
+}
+
+// Wait returns a channel that will receives tick.
+func (jt *PeriodicTimer) Wait() <-chan time.Time {
+	return jt.c
+}
+
+// Reset resets the current timer with the provided duration
+// NOTE: it is possible to receive a tick before the reset actually happen.
+func (jt *PeriodicTimer) Reset(d time.Duration) {
+	jt.resetOrDone <- d
+}
+
+// Stop stops the current timer but won't close the channel, this prevent a goroutine to received
+// a bad tick. This is the same strategy used by the time.Ticker.
+func (jt *PeriodicTimer) Stop() {
+	close(jt.resetOrDone)
+}
```

Comment:
This is why `(time.Timer).Stop` returns a bool. So to consume the signal on stop:

```
select {
case data := <-in:
  if timer.Stop() {
    // consumer spurious signal
    <-timer.C
  }
  process(data)
case <-timer.C:
  // operation timed out
  ...
}
```

By not consuming the signal the next operation could be marked as timed out immediately.

Reply:
OK, I've changed the code a bit, the Stop return a bool so you can consume the channel. Reset() now call Stop() and Start(). I have added tests to assert the scenario.